flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala

```diff
@@ -298,11 +333,34 @@ case class Join(
     val resolvedJoin = super.validate(tableEnv).asInstanceOf[Join]
     if (!resolvedJoin.condition.forall(_.resultType == BOOLEAN_TYPE_INFO)) {
       failValidation(s"filter expression ${resolvedJoin.condition} is not a boolean")
-    } else if (!ambiguousName.isEmpty) {
+    } else if (ambiguousName.nonEmpty) {
       failValidation(s"join relations with ambiguous names: ${ambiguousName.mkString(", ")}")
     }
+
+    resolvedJoin.condition.foreach(testJoinCondition(_))
     resolvedJoin
   }
+
+  private def testJoinCondition(expression: Expression): Unit = {
+    def checkIfJoinCondition(exp : Expression) = if (exp.children.exists(!_.isInstanceOf[JoinFieldReference])) {
+      failValidation(s"Only join predicates supported. For non-join predicates use Table#where.")
+    }
+
+    var equiJoinFound = false
+    def validateConditions(exp: Expression) : Unit = exp match {
```

Comment:
Checking whether it is an equi-join is a bit more difficult. For instance `l.a = r.b AND (l.c = r.d OR l.c = r.e)` is a valid equi-join condition. The first condition `(l.a = r.b)` can be pushed in the DataSet join (`left.join(right).where(a).equalTo(b)`) and the remaining condition can be evaluated in the `JoinFunction`.

To do the check, the predicate should be first normalized into "Conjunctive Normal Form" (CNF). Once that is done, the check is easy.